### PR TITLE
Improve texture scale & rotation

### LINKED ISSUE #322: Improve texture scale & rotation
### Description

* Rename Texture/`UV Rotation` to `Detiling`
* Add debug views for UV Scale/rotation
* Texture rotation doesn't properly rotate normals. It rotates them like an image rather than a 3D object. This video shows how https://www.youtube.com/watch?v=iXqCszt3LbE

---
Other items completed or tracked elsewhere

Brushing with tools should be able to change only certain features such as:
* ~~Paint UV scale / rotation, but don't change texture~~ Implemented in #365
* ~~Limit painting of color, texture, foliage to texture mask that are already present on the terrain(eg. inspector layers)~~ Duplicate #192

### PR DESCRIPTION
**Pull request by Xtarsia**
closes #322

### CODE REVIEW

**Comment by TokisanGames** [comment]
Great start, thank you. The debug views look great and the control rotation mode is fun to play with.

I noted a couple issues:

* Please change `uv_detiling` to just `detiling`.

* Normal rotation works for painted rotation, but not detiling. Some of the rotating segments do recalculate normals while moving, some do not.

Top Left: 45 deg, bottom left: 180 deg, right: 0 deg, detiling: 0:

image

Here it is with detiling at 0.5:

image

**Comment by Xtarsia** [comment]
Both issues should be resolved now!

**Comment by TokisanGames** [comment]
Yes, this looks great, thank you. 

I just noticed one other. It seems detiling and dual scaling have a little artifact. I don't notice it in near scale, only in far scale. Detile the rock and observe the hills down -X and you'll start seeing hard lines on the rotation squares. Move closer into near scale and the lines disappear.

image

File: src/shaders/main.glsl [hunk 1]
```diff
@@ -283,12 +292,13 @@ void get_material(vec2 base_uv, uint control, ivec3 iuv_center, vec3 normal, out
 	// be more optimal, the first introduces artifacts #276, and the second is noticably slower. 
 	// It seems the branching off dual scaling and the color array lookup is more optimal.
 	if (out_mat.blend > 0.f) {
+		// Unpack & rotate overlay normal for blending
+		normal_rg2.xz = unpack_normal(normal_rg2).xz;
+		normal_rg2.xz = rotate_normal(normal_rg2.xz,normal_angle2);
```
Comment (TokisanGames) [inline src/shaders/main.glsl:297]: `normal_rg2.xz, normal_angle2`
Please add spaces around operators like `*`, and spaces after commas

File: src/shaders/main.glsl [hunk 2]
```diff
@@ -221,11 +221,23 @@ vec4 height_blend(vec4 a_value, float a_height, vec4 b_value, float b_height, fl
...
+		normal_rotation += detile;
+	}
+	return uv;
+}
+
+vec2 rotate_normal(vec2 normal, float angle) {
+	angle += PI*0.5;
+	float new_y = dot(vec2(cos(angle),sin(angle)),normal);
+	angle -= PI*0.5;
+	float new_x = dot(vec2(cos(angle),sin(angle)),normal);
```
Comment (TokisanGames) [inline src/shaders/main.glsl:239]: `PI * 0.5`
Spaces after commas

File: src/shaders/main.glsl [hunk 3]
```diff
@@ -221,11 +221,23 @@ vec4 height_blend(vec4 a_value, float a_height, vec4 b_value, float b_height, fl
...
 
-vec2 uv_detiling(vec2 uv, vec2 uv_center, int mat_id){
-	vec2 rotation_center = floor(uv_center) + 0.5;
-	float random_angle_1 = (random(rotation_center) - 0.5) * 2.0 * TAU; // -180deg to 180deg
-	vec2 outUV = rotate_around(uv, rotation_center, random_angle_1 * _texture_uv_rotation_array[mat_id]);
-	return outUV;
+vec2 detiling(vec2 uv, vec2 uv_center, int mat_id, inout float normal_rotation){
+	if (_texture_detile_array[mat_id] >= 0.001){
+		uv_center = floor(uv_center)+0.5;
+		float detile = (random(uv_center)-0.5)* 2.0 * TAU * _texture_detile_array[mat_id]; // -180deg to 180deg
```
Comment (TokisanGames) [inline src/shaders/main.glsl:227]: ` + 0.5`
` - 0.5) * 2.0`

File: src/shaders/dual_scaling.glsl [hunk 4]
```diff
@@ -36,12 +46,16 @@ uniform float dual_scale_near : hint_range(0,1000) = 100.0;
...
-	matUV = uv_detiling(base_uv * mat_scale, uv_center * mat_scale, out_mat.base);
+	matUV = detiling(base_uv * mat_scale, uv_center * mat_scale, out_mat.base, normal_angle);
 	ddx1 *= mat_scale;
 	ddy1 *= mat_scale;
 	albedo_ht = textureGrad(_texture_array_albedo, vec3(matUV, float(out_mat.base)), ddx1, ddy1);
 	normal_rg = textureGrad(_texture_array_normal, vec3(matUV, float(out_mat.base)), ddx1, ddy1);
 
+	// Unpack & rotate base normal for blending
+	normal_rg.xz = unpack_normal(normal_rg).xz;
+	normal_rg.xz = rotate_normal(normal_rg.xz,normal_angle);
```
Comment (TokisanGames) [inline src/shaders/dual_scaling.glsl:57]: Space after comma

File: src/shaders/dual_scaling.glsl [hunk 5]
```diff
@@ -15,18 +15,28 @@ uniform float dual_scale_near : hint_range(0,1000) = 100.0;
...
 		ddy1 *= dual_scale_reduction;
-		matUV = uv_detiling(base_uv * mat_scale, uv_center * mat_scale, dual_scale_texture);
+		float dual_scale_normal = uv_rotation; //do not add near & far rotations
+		matUV = detiling(base_uv * mat_scale, uv_center * mat_scale, dual_scale_texture, dual_scale_normal);
 		albedo_far = textureGrad(_texture_array_albedo, vec3(matUV, float(dual_scale_texture)), ddx1, ddy1);
 		normal_far = textureGrad(_texture_array_normal, vec3(matUV, float(dual_scale_texture)), ddx1, ddy1);
+
+		// Unpack & rotate dual scale normal for blending
+		normal_far.xz = unpack_normal(normal_far).xz;
+		normal_far.xz = rotate_normal(normal_far.xz,dual_scale_normal);
```
Comment (TokisanGames) [inline src/shaders/dual_scaling.glsl:39]: Space

File: src/shaders/dual_scaling.glsl [hunk 6]
```diff
@@ -15,18 +15,28 @@ uniform float dual_scale_near : hint_range(0,1000) = 100.0;
...
-	matUV = uv_detiling(base_uv * mat_scale, uv_center * mat_scale, out_mat.base);
+	matUV = detiling(base_uv * mat_scale, uv_center * mat_scale, out_mat.base, normal_angle);
 	ddx1 *= mat_scale;
 	ddy1 *= mat_scale;
 	albedo_ht = textureGrad(_texture_array_albedo, vec3(matUV, float(out_mat.base)), ddx1, ddy1);
 	normal_rg = textureGrad(_texture_array_normal, vec3(matUV, float(out_mat.base)), ddx1, ddy1);
+
+	// Unpack & rotate base normal for blending
+	normal_rg.xz = unpack_normal(normal_rg).xz;
+	normal_rg.xz = rotate_normal(normal_rg.xz,normal_angle);
```
Comment (TokisanGames) [inline src/shaders/dual_scaling.glsl:26]: Space

File: src/shaders/debug_views.glsl [hunk 7]
```diff
@@ -61,6 +61,34 @@ R"(
...
+	uint __s_control = texelFetch(_control_maps, __suv, 0).r;
+	uint __scale = (__s_control >>7u & 0x7u);
+	vec3 __s_colors[8] = {
+		vec3(.5,.5,.5),vec3(.675,.25,.375),vec3(.75,.125,.25),vec3(.875,.0,.125),vec3(1.,0.,0.),
+		vec3(0.,0.,1.),vec3(.0,.166,.833),vec3(.166,.333,.666)
+	};
+	ALBEDO = __s_colors[__scale];
+	ROUGHNESS = 1.;
+	SPECULAR = 0.;
+	NORMAL_MAP = vec3(0.5,0.5,1.0);
```
Comment (TokisanGames) [inline src/shaders/debug_views.glsl:90]: Comma spacing all throughout this block.

**Comment by Xtarsia** [comment]
hmmm so fixing detiling with dual scaling (mostly tri-scaling) is problematic, since the bigger the area the texture covers reletive to the vertex grid, the less the bilerp works to cover edge blend between rotated UV tiles.

either, do additional reads for a manual blend, or disable detiling when dual scaling. (or at least tri-scaling)

File: src/shaders/debug_views.glsl (hunk 7 above)
Comment (Xtarsia) [reply src/shaders/debug_views.glsl]: resolved in 3a4139a

File: src/shaders/dual_scaling.glsl (hunk 6 above)
Comment (Xtarsia) [reply src/shaders/dual_scaling.glsl]: resolved in 3a4139a

File: src/shaders/dual_scaling.glsl (hunk 5 above)
Comment (Xtarsia) [reply src/shaders/dual_scaling.glsl]: resolved in 3a4139a

File: src/shaders/dual_scaling.glsl (hunk 4 above)
Comment (Xtarsia) [reply src/shaders/dual_scaling.glsl]: resolved in 3a4139a

File: src/shaders/main.glsl (hunk 3 above)
Comment (Xtarsia) [reply src/shaders/main.glsl]: resolved in 3a4139a

File: src/shaders/main.glsl (hunk 2 above)
Comment (Xtarsia) [reply src/shaders/main.glsl]: resolved in 3a4139a

File: src/shaders/main.glsl (hunk 1 above)
Comment (Xtarsia) [reply src/shaders/main.glsl]: resolved in 3a4139a

**Comment by Xtarsia** [comment]
a different approach, is the mix only the rgb values when dual scaling and leave the height value untouch. this also makes the transition much less noticable when moveing around the world as the height blend values dont change.

`	float far_factor = clamp(smoothstep(dual_scale_near, dual_scale_far, length(v_vertex - v_camera_pos)), 0.0, 1.0);
	if(out_mat.base == dual_scale_texture) {
		albedo_ht.rgb = mix(albedo_ht.rgb, albedo_far.rgb, far_factor);
		//albedo_ht = mix(albedo_ht, albedo_far, far_factor);
		normal_rg = mix(normal_rg, normal_far, far_factor);
	}`

it doesnt completley solve the above issue but may be worth considering anyways

File: src/shaders/dual_scaling.glsl [hunk 8]
```diff
@@ -30,7 +30,8 @@ uniform float dual_scale_near : hint_range(0,1000) = 100.0;
 		ddx1 *= dual_scale_reduction;
 		ddy1 *= dual_scale_reduction;
 		float dual_scale_normal = uv_rotation; //do not add near & far rotations
-		matUV = detiling(base_uv * mat_scale, uv_center * mat_scale, dual_scale_texture, dual_scale_normal);
+		// Do not apply detiling if tri-scale reduction occurs.
+		matUV = region < 0 ? base_uv * mat_scale : detiling(base_uv * mat_scale, uv_center * mat_scale, dual_scale_texture, dual_scale_normal);
```
Comment (TokisanGames) [inline src/shaders/dual_scaling.glsl:34]: This is an acceptable fix for now and we can look at it again when considering multiscaling.

**Comment by TokisanGames** [comment]
This branch is broken. Github won't update after a force push. Most likely due to the ampersand in the name.

**Pull request closed without merging**